Sources/CasePaths/EnumReflection.swift

```diff
@@ -93,57 +110,240 @@ public func extract<Root, Value>(_ embed: @escaping (Value) -> Root) -> (Root) -
+      swift_release(pair.heapObject)
+      return value
+    }
+
+    return answer
+  }
+
+  func allocateBoxForExistential(in buffer: UnsafeMutableRawPointer) -> UnsafeMutableRawPointer {
+    guard valueWitnessTable.flags.contains(.isNonInline) else {
+      return buffer
+    }
+
+    let (heapObject: heapObject, buffer: boxBuffer) = swift_allocBox(for: ptr)
+    buffer.storeBytes(of: heapObject, as: UnsafeMutableRawPointer.self)
+    return boxBuffer
+  }
+
+  func deallocateBoxForExistential(in buffer: UnsafeMutableRawPointer) {
+    guard valueWitnessTable.flags.contains(.isNonInline) else { return }
+    swift_deallocBox(buffer.load(as: UnsafeMutableRawPointer.self))
+  }
+}
+
+@_silgen_name("swift_allocBox")
+private func swift_allocBox(for metadata: UnsafeRawPointer) -> BoxPair
+
+@_silgen_name("swift_allocBox")
+private func swift_deallocBox(_ heapObject: UnsafeMutableRawPointer)
+
+@_silgen_name("swift_projectBox")
+private func swift_projectBox(_ heapObject: UnsafeMutableRawPointer) -> UnsafeMutableRawPointer
+
+@_silgen_name("swift_release")
+private func swift_release(_ heapObject: UnsafeMutableRawPointer)
+
 private struct EnumTypeDescriptor {
-  // These fields are not modeled because we don't need them.
-  // They are the type descriptor flags and various pointer offsets.
-  let flags, p1, p2, p3, p4: Int32
+  let ptr: UnsafeRawPointer
 
-  let numPayloadCasesAndPayloadSizeOffset: Int32
-  let numEmptyCases: Int32
+  var fieldDescriptor: FieldDescriptor {
+    return FieldDescriptor(
+      ptr: ptr.advanced(by: 4 * 4).loadRelativePointer())
+  }
 
+  #if compiler(<5.2)
   var numPayloadCases: Int32 {
-    numPayloadCasesAndPayloadSizeOffset & 0xFFFFFF
+    return ptr.advanced(by: 5 * 4).load(as: Int32.self) & 0xFFFFFF
   }
+
+  var numEmptyCases: Int32 {
+    return ptr.advanced(by: 6 * 4).load(as: Int32.self)
+  }
+
+  var isInhabited: Bool { numPayloadCases + numEmptyCases > 0 }
+  #endif
 }
 
-private func isUninhabitedEnum(_ type: Any.Type) -> Bool {
-  // Load the type kind from the common type metadata area. Memory layout reference:
-  // https://github.com/apple/swift/blob/master/docs/ABI/TypeMetadata.rst
-  let metadataPtr = unsafeBitCast(type, to: UnsafeRawPointer.self)
-  let metadataKind = metadataPtr.load(as: Int.self)
+private struct FieldDescriptor {
+  let ptr: UnsafeRawPointer
 
-  // Check that this is an enum. Value reference:
-  // https://github.com/apple/swift/blob/master/stdlib/public/core/ReflectionMirror.swift
-  let isEnum = metadataKind == 0x201
-  guard isEnum else { return false }
+  /// The size of a FieldRecord as stored in the executable.
+  var recordSize: Int { Int(ptr.advanced(by: 2 * 4 + 2).load(as: UInt16.self)) }
 
-  // Access enum type descriptor
-  let enumMetadata = metadataPtr.load(as: EnumMetadata.self)
-  let enumTypeDescriptor = enumMetadata.typeDescriptor.pointee
+  var fieldCount: UInt32 { ptr.advanced(by: 2 * 4 + 2 * 2).load(as: UInt32.self) }
```

Comment:
Looks like this isn't used?

Reply:
Indeed it is not used because I just assume the enum tag is a valid index into the FieldRecords. The C++ code also makes this assumption.